### PR TITLE
Support loss dynamics tracking for CustomSingleStageDetector

### PR DESCRIPTION
**Pull request by vinnamkim**
### Summary
- Ticket no. 93287
- Add training loss dynamics exportation feature for `CustomSingleStageDetector` model.

### How to test

I added unit tests about this change.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py [hunk 1]
```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
...
+    def _get_pos_inds(self, labels):
+        pos_inds = super()._get_pos_inds(labels)
+
+        if len(pos_inds) > 0:
+            pos_assigned_gt_inds = self.all_pos_assigned_gt_inds[self.cur_loss_idx].reshape(-1)
+
+            gt_inds = pos_assigned_gt_inds[pos_inds].cpu()
+
+            self.batch_inds = gt_inds // self.max_gt_bboxes_len
+            self.bbox_inds = gt_inds % self.max_gt_bboxes_len
```
Comment (wonjuleee) [inline otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py]: should this be done at here? We may be possible to reuse `CustomSSDHead`'s one if we move this if statement.

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py [hunk 2]
```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
...
+
+        self.pos_inds = pos_inds
+        return pos_inds
+
+    def _get_loss_cls(self, num_total_samples, loss_cls_all, pos_inds, topk_loss_cls_neg):
+        loss_cls_pos = loss_cls_all[pos_inds]
+        loss_cls_neg = topk_loss_cls_neg.sum()
+        loss_cls = (loss_cls_pos.sum() + loss_cls_neg) / num_total_samples
+
+        self._store_loss_dyns(loss_cls_pos.detach(), TrackingLossType.cls)
```
Comment (wonjuleee) [inline otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:219]: If we move this to loss function for `TrackingLossDynamics`, we may be reuse the `CustomSSDHead`'s one too.

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py [hunk 3]
```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
...
+        loss_bbox = smooth_l1_loss(
+            bbox_pred,
+            bbox_targets,
+            bbox_weights,
+            beta=self.train_cfg.smoothl1_beta,
+            avg_factor=num_total_samples,
+            reduction="none",
+        )
+
+        self._store_loss_dyns(loss_bbox[self.pos_inds].detach().mean(-1), TrackingLossType.bbox)
```
Comment (wonjuleee) [inline otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:232]: here also.

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py [hunk 4]
```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
...
+            label_channels,
+            unmap_outputs,
+            return_sampling_results,
+        )
+
+        all_pos_assigned_gt_inds = images_to_levels(self.pos_assigned_gt_inds_list, num_level_anchors)
+        self.all_pos_assigned_gt_inds = torch.cat(all_pos_assigned_gt_inds, -1)
+        return targets
+
+    def _get_targets_single(
```
Comment (wonjuleee) [inline otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:261]: why this should be required?

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py [hunk 5]
```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
...
+        gt_labels_list=None,
+        label_channels=1,
+        unmap_outputs=True,
+        return_sampling_results=False,
+    ):
+        """Get targets for Detection head."""
+        num_level_anchors = [anchors.size(0) for anchors in anchor_list[0]]
+        self.max_gt_bboxes_len = max([len(gt_bboxes) for gt_bboxes in gt_bboxes_list])
+        self.cur_batch_idx = 0
+        self.pos_assigned_gt_inds_list = []
```
Comment (wonjuleee) [inline otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py]: Above looks like independent to call `super().get_targets()`. Is this any other way to reuse `get_targets` of `super` by taking out them from this function?

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py (hunk 2 above)
Comment (vinnamkim) [reply otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:219]: https://github.com/openvinotoolkit/training_extensions/pull/2115/commits/cbf7ecc05780fc23d5d854a59942beb0d401f539 tries to move the shared codes between ATSS and SSD to `class TrackingLossDynamicsMixIn` as possible:
https://github.com/openvinotoolkit/training_extensions/blob/cbf7ecc05780fc23d5d854a59942beb0d401f539/otx/algorithms/detection/adapters/mmdet/models/heads/cross_dataset_detector_head.py#L283-L289

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py (hunk 3 above)
Comment (vinnamkim) [reply otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:232]: Please refer to my comment above.

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py (hunk 5 above)
Comment (vinnamkim) [reply otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py]: Please refer to my comment about code sharing of https://github.com/openvinotoolkit/training_extensions/pull/2115/commits/cbf7ecc05780fc23d5d854a59942beb0d401f539.

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py (hunk 1 above)
Comment (vinnamkim) [reply otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py]: https://github.com/openvinotoolkit/training_extensions/pull/2115/commits/cbf7ecc05780fc23d5d854a59942beb0d401f539 moves those lines into the comman mix-in class:
https://github.com/openvinotoolkit/training_extensions/blob/cbf7ecc05780fc23d5d854a59942beb0d401f539/otx/algorithms/detection/adapters/mmdet/models/heads/cross_dataset_detector_head.py#L269-L281

File: otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py (hunk 4 above)
Comment (vinnamkim) [reply otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py:261]: This is because we should store `sampling_result.pos_assigned_gt_inds` to match the loss dynamics statistics obtained from the anchors with the source bounding boxes for tracking.
https://github.com/openvinotoolkit/training_extensions/blob/cbf7ecc05780fc23d5d854a59942beb0d401f539/otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py#L347-L359

**Review by wonjuleee — DISMISSED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by sungmanc — DISMISSED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by wonjuleee — APPROVED** [review]
